# AgentOS Rules Authoring Guideï¼ˆè§„åˆ™ç¼–å†™æŒ‡å—ï¼‰

æœ¬æŒ‡å—å¸®åŠ©ä½ åˆ›å»ºç¬¦åˆ v0.9 æ ‡å‡†çš„ AgentOS Rulesã€‚

---

## ğŸ¯ Rule 0ï¼šéµå®ˆäº”æ¡çº¢çº¿

åœ¨ç¼–å†™ä»»ä½• rule å‰ï¼Œå¿…é¡»ç†è§£å¹¶éµå®ˆä»¥ä¸‹çº¢çº¿ï¼š

1. **RL1**: Rule ä¸åŒ…å«æ‰§è¡ŒæŒ‡ä»¤
2. **RL2**: Rule å¿…é¡»å¯å®¡è®¡ï¼ˆevidence_requiredï¼‰
3. **RL3**: Rule å¿…é¡»å¯æœºå™¨åˆ¤å®šï¼ˆpredicate ç»“æ„åŒ–ï¼‰
4. **RL4**: Rule å¿…é¡»å£°æ˜é€‚ç”¨èŒƒå›´ï¼ˆscopeï¼‰
5. **RL5**: Rule å¿…é¡»æœ‰ lineage

è¯¦è§ `README.md` çº¢çº¿ç« èŠ‚ã€‚

---

## ğŸ“‹ æ ‡å‡† Rule ç»“æ„

### æœ€å°æ¨¡æ¿

```yaml
id: rule_r<nn>_<short_name>
type: rule
title: "<Human-readable title>"
description: "<Detailed description of what this rule enforces>"
version: "0.9.0"
status: active

lineage:
  introduced_in: "v0.9"
  derived_from: null
  supersedes: []

constraints:
  execution: forbidden

rule:
  severity: error  # info|warn|error|block
  scope:
    applies_to_types: ["workflow", "agent", "command"]
    applies_to_risk: ["medium", "high"]
    applies_to_phases: ["implementation", "review"]
  
  when:
    <structured_condition>: true
  
  then:
    decision: deny  # allow|deny|warn|require_review
    reason: "<Human-readable reason>"
    required_changes:
      - "<Change 1>"
      - "<Change 2>"
  
  evidence_required:
    - "<evidence_type_1>"
    - "<evidence_type_2>"

metadata:
  tags: ["tag1", "tag2"]
  author: "AgentOS Team"
```

### å®Œæ•´æ¨¡æ¿

```yaml
id: rule_r<nn>_<short_name>
type: rule
title: "<Human-readable title (5-200 chars)>"
description: >
  <Detailed description of what this rule enforces.
  Explain the rationale, use cases, and consequences.
  (10-2000 chars)>
version: "0.9.0"
status: active  # draft|active|deprecated

lineage:
  introduced_in: "v0.9"      # é¦–æ¬¡å¼•å…¥ç‰ˆæœ¬
  derived_from: null         # çˆ¶ rule IDï¼ˆroot ä¸º nullï¼‰
  supersedes: []             # æ›¿ä»£çš„æ—§ rule IDs

constraints:
  execution: forbidden       # ğŸš¨ RL1: å¿…é¡»ä¸º forbidden

rule:
  severity: error            # ğŸš¨ RL3: info|warn|error|block
  
  scope:                     # ğŸš¨ RL4: è‡³å°‘ä¸€ä¸ªéç©º
    applies_to_types:        # workflow|agent|command|policy|memory|fact|rule
      - "workflow"
      - "command"
    applies_to_risk:         # low|medium|high|critical
      - "high"
      - "critical"
    applies_to_phases:       # setup|analysis|...|postmortem
      - "implementation"
      - "review"
  
  when:                      # ğŸš¨ RL3: ç»“æ„åŒ–æ¡ä»¶ï¼ˆä¸èƒ½æ˜¯å­—ç¬¦ä¸²ï¼‰
    any_of:                  # æˆ– all_of
      - field_exists: "forbidden_field"
      - field_missing: "required_field"
      - custom_condition: true
  
  then:                      # ğŸš¨ RL3: ç»“æ„åŒ–å†³ç­–ï¼ˆä¸èƒ½æ˜¯å­—ç¬¦ä¸²ï¼‰
    decision: deny           # allow|deny|warn|require_review
    reason: "Human-readable reason for this decision"
    required_changes:        # å¯é€‰
      - "Change description 1"
      - "Change description 2"
  
  evidence_required:         # ğŸš¨ RL2: éç©ºæ•°ç»„
    - "content_source_yaml"
    - "schema_validation"
    - "factpack"

metadata:                    # å¯é€‰
  tags:
    - "category1"
    - "category2"
  related_rules:
    - "rule_r01_no_execution"
  documentation_url: "https://example.com/docs"
  author: "Your Name"
  examples:
    - "Example use case 1"
```

---

## ğŸ”§ å­—æ®µè¯¦è§£

### å¿…éœ€å­—æ®µ

#### `id`
- **æ ¼å¼**: `rule_r<nn>_<short_name>`
  - `<nn>`: ä¸¤ä½æ•°å­—ï¼ˆ01-99ï¼‰
  - `<short_name>`: ä¸‹åˆ’çº¿åˆ†éš”çš„çŸ­åç§°
- **ç¤ºä¾‹**: `rule_r01_no_execution`, `rule_r12_rollback_plan_required_high_risk`
- **çº¦æŸ**: å¿…é¡»å”¯ä¸€ï¼Œå¿…é¡»ä¸æ–‡ä»¶ååŒ¹é…ï¼ˆ`<id>.yaml`ï¼‰

#### `type`
- **å›ºå®šå€¼**: `"rule"`

#### `version`
- **æ ¼å¼**: `<major>.<minor>.<patch>`
- **ç¤ºä¾‹**: `"0.9.0"`
- **è¯´æ˜**: v0.9 é¦–æ¬¡å¼•å…¥çš„ rules éƒ½æ˜¯ `0.9.0`

#### `title`
- **é•¿åº¦**: 5-200 å­—ç¬¦
- **è¦æ±‚**: ç®€æ´ã€æè¿°æ€§å¼º
- **ç¤ºä¾‹**: `"No execution is allowed in content plane"`

#### `description`
- **é•¿åº¦**: 10-2000 å­—ç¬¦
- **è¦æ±‚**: è¯¦ç»†è¯´æ˜è§„åˆ™çš„ç”¨é€”ã€åŸå› ã€å½±å“
- **å»ºè®®**: åŒ…å«é€‚ç”¨åœºæ™¯å’Œä¸é€‚ç”¨åœºæ™¯

#### `status`
- **æšä¸¾**: `draft`, `active`, `deprecated`
- **è¯´æ˜**:
  - `draft`: å¼€å‘ä¸­ï¼Œæœªæ­£å¼ä½¿ç”¨
  - `active`: æ­£å¼å¯ç”¨
  - `deprecated`: å·²åºŸå¼ƒï¼Œè¢«æ–°è§„åˆ™æ›¿ä»£

#### `lineage`ï¼ˆğŸš¨ RL5ï¼‰
- **å¿…é¡»å­—æ®µ**: `introduced_in`, `derived_from`, `supersedes`
- **ç¤ºä¾‹**:
  ```yaml
  lineage:
    introduced_in: "v0.9"     # æ ¼å¼: v<major>.<minor>
    derived_from: null         # root ç‰ˆæœ¬ä¸º null
    supersedes: []             # æ›¿ä»£çš„æ—§è§„åˆ™ï¼ˆå¯ä¸ºç©ºæ•°ç»„ï¼‰
  ```

#### `constraints`ï¼ˆğŸš¨ RL1ï¼‰
- **å¿…é¡»å­—æ®µ**: `execution`
- **å›ºå®šå€¼**: `execution: forbidden`
- **è¯´æ˜**: å¼ºåˆ¶ rules ä¸åŒ…å«æ‰§è¡Œé€»è¾‘

#### `rule`
æ•´ä¸ª rule çš„æ ¸å¿ƒé€»è¾‘ï¼ŒåŒ…å« 5 ä¸ªå¿…éœ€å­å­—æ®µï¼š

##### `rule.severity`ï¼ˆğŸš¨ RL3ï¼‰
- **æšä¸¾**: `info`, `warn`, `error`, `block`
- **è¯´æ˜**:
  - `info`: ä¿¡æ¯æç¤ºï¼Œä¸å½±å“æµç¨‹
  - `warn`: è­¦å‘Šï¼Œè®°å½•ä½†å…è®¸é€šè¿‡
  - `error`: é”™è¯¯ï¼Œé»˜è®¤æ‹’ç»ï¼ˆéœ€ä¿®æ­£ï¼‰
  - `block`: é˜»å¡ï¼Œæ— æ³•é€šè¿‡ï¼ˆçº¢çº¿çº§åˆ«ï¼‰

##### `rule.scope`ï¼ˆğŸš¨ RL4ï¼‰
è‡³å°‘ä¸€ä¸ªå­å­—æ®µéç©ºï¼š
- `applies_to_types`: content ç±»å‹æ•°ç»„
  - å¯é€‰å€¼: `workflow`, `agent`, `command`, `policy`, `memory`, `fact`, `rule`
- `applies_to_risk`: é£é™©çº§åˆ«æ•°ç»„
  - å¯é€‰å€¼: `low`, `medium`, `high`, `critical`
- `applies_to_phases`: å·¥ä½œæµé˜¶æ®µæ•°ç»„
  - å¯é€‰å€¼: `setup`, `analysis`, `design`, `implementation`, `validation`, `review`, `release`, `operations`, `postmortem`

##### `rule.when`ï¼ˆğŸš¨ RL3ï¼‰
- **ç±»å‹**: objectï¼ˆç»“æ„åŒ–æ¡ä»¶ï¼‰
- **ç¦æ­¢**: å­—ç¬¦ä¸²ï¼ˆå¦‚ `when: "if risk is high"`ï¼‰
- **æ¨èæ¨¡å¼**:
  ```yaml
  when:
    any_of:          # æˆ– all_of
      - field_exists: "execute"
      - field_missing: "lineage"
      - risk_level_high: true
  ```

##### `rule.then`ï¼ˆğŸš¨ RL3ï¼‰
- **ç±»å‹**: objectï¼ˆç»“æ„åŒ–å†³ç­–ï¼‰
- **å¿…é¡»å­—æ®µ**: `decision`
- **decision æšä¸¾**: `allow`, `deny`, `warn`, `require_review`
- **å¯é€‰å­—æ®µ**: `reason`, `required_changes`
- **ç¤ºä¾‹**:
  ```yaml
  then:
    decision: deny
    reason: "Execution payload is forbidden"
    required_changes:
      - "Remove execute field"
  ```

##### `rule.evidence_required`ï¼ˆğŸš¨ RL2ï¼‰
- **ç±»å‹**: arrayï¼ˆéç©ºï¼‰
- **è¯´æ˜**: åˆ¤å®šè§„åˆ™éœ€è¦å“ªäº›è¯æ®
- **å¸¸è§å€¼**:
  - `content_source_yaml`: YAML æºæ–‡ä»¶
  - `schema_validation`: Schema éªŒè¯ç»“æœ
  - `factpack`: Factpack æ‰«æç»“æœ
  - `project_scan`: é¡¹ç›®æ‰«æç»“æœ
  - `registry_query_result`: Registry æŸ¥è¯¢
  - `risk_assessment`: é£é™©è¯„ä¼°
  - `audit_log_entry`: å®¡è®¡æ—¥å¿—

---

## ğŸ“ ç¼–å†™æ­¥éª¤

### Step 1: ç¡®å®šè§„åˆ™ç›®çš„

å›ç­”ä»¥ä¸‹é—®é¢˜ï¼š
1. è¿™æ¡è§„åˆ™è¦é˜²æ­¢ä»€ä¹ˆé—®é¢˜ï¼Ÿ
2. é€‚ç”¨äºå“ªäº› content types / risk levels / phasesï¼Ÿ
3. è¿åè§„åˆ™çš„åæœæ˜¯ä»€ä¹ˆï¼Ÿ
4. éœ€è¦å“ªäº›è¯æ®æ¥åˆ¤å®šï¼Ÿ

### Step 2: åˆ†é… ID å’Œå‘½å

- æŸ¥çœ‹ `catalog.md` ç¡®å®šä¸‹ä¸€ä¸ªå¯ç”¨ç¼–å·ï¼ˆå¦‚ R13ï¼‰
- èµ·ä¸€ä¸ªæè¿°æ€§çš„çŸ­åç§°ï¼ˆå¦‚ `rollback_required`ï¼‰
- ç»„åˆæˆ IDï¼š`rule_r13_rollback_required`

### Step 3: å¡«å†™åŸºæœ¬å­—æ®µ

```yaml
id: rule_r13_rollback_required
type: rule
title: "Rollback plan required for destructive operations"
description: >
  All commands with destructive effects (delete, drop) must include
  a rollback_plan to undo changes if execution fails.
version: "0.9.0"
status: draft  # å…ˆç”¨ draftï¼Œæµ‹è¯•é€šè¿‡åæ”¹ä¸º active
```

### Step 4: å®šä¹‰ lineage

```yaml
lineage:
  introduced_in: "v0.9"
  derived_from: null  # æ–°è§„åˆ™ï¼Œæ— çˆ¶ç‰ˆæœ¬
  supersedes: []      # ä¸æ›¿ä»£å…¶ä»–è§„åˆ™
```

### Step 5: è®¾ç½® constraints

```yaml
constraints:
  execution: forbidden  # å›ºå®šå€¼
```

### Step 6: ç¼–å†™ rule é€»è¾‘

```yaml
rule:
  severity: error  # é€‰æ‹©åˆé€‚çš„çº§åˆ«
  
  scope:  # è‡³å°‘ä¸€ä¸ªéç©º
    applies_to_types: ["command"]
    applies_to_risk: ["high", "critical"]
    applies_to_phases: ["implementation", "review"]
  
  when:  # ç»“æ„åŒ–æ¡ä»¶
    all_of:
      - effect_kind_is_delete: true
      - rollback_plan_missing: true
  
  then:  # ç»“æ„åŒ–å†³ç­–
    decision: deny
    reason: "Destructive commands must have rollback plan"
    required_changes:
      - "Add rollback_plan field"
      - "Specify undo steps"
  
  evidence_required:  # éç©º
    - "command_effects"
    - "rollback_plan"
```

### Step 7: æ·»åŠ  metadataï¼ˆå¯é€‰ï¼‰

```yaml
metadata:
  tags: ["rollback", "safety", "destructive"]
  related_rules: ["rule_r12_rollback_plan_required_high_risk"]
  author: "Your Name"
```

### Step 8: ä¿å­˜æ–‡ä»¶

- æ–‡ä»¶åå¿…é¡»ä¸ ID åŒ¹é…ï¼š`rule_r13_rollback_required.yaml`
- è·¯å¾„ï¼š`docs/content/rules/p0/rule_r13_rollback_required.yaml`

---

## âœ… éªŒè¯æ¸…å•

æäº¤å‰æ£€æŸ¥ï¼š

### çº¢çº¿æ£€æŸ¥
- [ ] `constraints.execution` æ˜¯ `forbidden`ï¼ˆRL1ï¼‰
- [ ] `rule.evidence_required` éç©ºæ•°ç»„ï¼ˆRL2ï¼‰
- [ ] `rule.when` å’Œ `rule.then` æ˜¯ç»“æ„åŒ–å¯¹è±¡ï¼ˆRL3ï¼‰
- [ ] `rule.severity` æ˜¯ info/warn/error/blockï¼ˆRL3ï¼‰
- [ ] `rule.scope` è‡³å°‘ä¸€ä¸ªå­—æ®µéç©ºï¼ˆRL4ï¼‰
- [ ] `lineage` åŒ…å« introduced_in/derived_from/supersedesï¼ˆRL5ï¼‰

### ç»“æ„æ£€æŸ¥
- [ ] ID æ ¼å¼æ­£ç¡®ï¼ˆ`rule_r<nn>_<name>`ï¼‰
- [ ] æ–‡ä»¶åä¸ ID åŒ¹é…ï¼ˆ`<id>.yaml`ï¼‰
- [ ] `type` æ˜¯ `"rule"`
- [ ] `version` æ ¼å¼æ­£ç¡®ï¼ˆ`X.Y.Z`ï¼‰
- [ ] `title` é•¿åº¦ 5-200 å­—ç¬¦
- [ ] `description` é•¿åº¦ 10-2000 å­—ç¬¦

### é€»è¾‘æ£€æŸ¥
- [ ] `when` æ¡ä»¶æ¸…æ™°ã€å¯éªŒè¯
- [ ] `then.decision` æ˜¯æœ‰æ•ˆæšä¸¾å€¼
- [ ] `scope` èŒƒå›´åˆç†
- [ ] `evidence_required` ç±»å‹æ˜ç¡®

---

## ğŸ§ª æœ¬åœ°æµ‹è¯•

### 1. Schema éªŒè¯

```bash
uv run python scripts/convert_rules.py --validate --file p0/rule_r<nn>_<name>.yaml
```

**é¢„æœŸè¾“å‡º**ï¼š
```
âœ… Validation passed
```

### 2. çº¢çº¿éªŒè¯

```bash
uv run python scripts/register_rules.py --validate-only
```

**é¢„æœŸè¾“å‡º**ï¼š
```
âœ… rule_r<nn>_<name>: All red lines passed
```

### 3. æ³¨å†Œæµ‹è¯•

```bash
# ä½¿ç”¨ä¸´æ—¶æ•°æ®åº“æµ‹è¯•
uv run python scripts/register_rules.py --db /tmp/test_rules.db
```

**é¢„æœŸè¾“å‡º**ï¼š
```
âœ… Registered: rule_r<nn>_<name> v0.9.0
```

---

## ğŸš« å¸¸è§é”™è¯¯

### é”™è¯¯ 1: when/then ç”¨å­—ç¬¦ä¸²

```yaml
# âŒ é”™è¯¯
rule:
  when: "if risk is high"
  then: "deny it"
```

**æ­£ç¡®åšæ³•**ï¼š
```yaml
# âœ… æ­£ç¡®
rule:
  when:
    risk_level_high: true
  then:
    decision: deny
    reason: "High risk not allowed"
```

---

### é”™è¯¯ 2: evidence_required ä¸ºç©º

```yaml
# âŒ é”™è¯¯
rule:
  evidence_required: []
```

**æ­£ç¡®åšæ³•**ï¼š
```yaml
# âœ… æ­£ç¡®
rule:
  evidence_required:
    - "content_source_yaml"
```

---

### é”™è¯¯ 3: scope å…¨éƒ¨ä¸ºç©º

```yaml
# âŒ é”™è¯¯
rule:
  scope:
    applies_to_types: []
    applies_to_risk: []
    applies_to_phases: []
```

**æ­£ç¡®åšæ³•**ï¼š
```yaml
# âœ… æ­£ç¡®
rule:
  scope:
    applies_to_types: ["command"]
    # è‡³å°‘ä¸€ä¸ªéç©ºå³å¯
```

---

### é”™è¯¯ 4: åŒ…å«æ‰§è¡Œå­—æ®µ

```yaml
# âŒ é”™è¯¯
rule:
  when: { ... }
  then:
    execute: "rm -rf /tmp"  # è¿å RL1
```

**æ­£ç¡®åšæ³•**ï¼š
```yaml
# âœ… æ­£ç¡®
rule:
  when: { ... }
  then:
    decision: deny  # åªèƒ½æ˜¯ gate decision
    reason: "..."
```

---

### é”™è¯¯ 5: æ–‡ä»¶åä¸ ID ä¸åŒ¹é…

```yaml
# âŒ é”™è¯¯
# æ–‡ä»¶: rule_r13_wrong_name.yaml
id: rule_r13_rollback_required
```

**æ­£ç¡®åšæ³•**ï¼š
```yaml
# âœ… æ­£ç¡®
# æ–‡ä»¶: rule_r13_rollback_required.yaml
id: rule_r13_rollback_required
```

---

## ğŸ“– æœ€ä½³å®è·µ

### 1. æ¸…æ™°å‘½å

- **å¥½**: `rule_r01_no_execution`
- **å·®**: `rule_r01_rule1`

### 2. å…·ä½“æè¿°

- **å¥½**: "Commands with write effects must enter requires_review state"
- **å·®**: "Write commands need review"

### 3. ç»“æ„åŒ–æ¡ä»¶

- **å¥½**: `{ effect_kind_is_write: true }`
- **å·®**: `"if effect is write"`

### 4. åˆç† severity

- `block`: çº¢çº¿çº§åˆ«ï¼ˆå¦‚ R01 ç¦æ­¢æ‰§è¡Œï¼‰
- `error`: å¿…é¡»éµå®ˆï¼ˆå¦‚ R03 å¼•ç”¨å¿…é¡»åœ¨ registryï¼‰
- `warn`: æœ€ä½³å®è·µï¼ˆå¦‚ R09 å»ºè®®å¼•ç”¨è¯æ®ï¼‰
- `info`: æç¤ºä¿¡æ¯ï¼ˆç½•ç”¨ï¼‰

### 5. æ˜ç¡® scope

ä¸è¦ç”¨ `applies_to_types: ["workflow", "agent", "command", "policy", "memory", "fact", "rule"]`ï¼ˆå¤ªå®½æ³›ï¼‰

**å»ºè®®**: åªåˆ—å‡ºçœŸæ­£é€‚ç”¨çš„ç±»å‹

---

## ğŸ”„ æ¼”åŒ–è§„åˆ™

### åˆ›å»ºæ–°ç‰ˆæœ¬

å¦‚æœéœ€è¦ä¿®æ”¹ç°æœ‰è§„åˆ™ï¼š

```yaml
id: rule_r01_no_execution_v2  # æ–° ID
type: rule
title: "..."
version: "0.10.0"  # æ–°ç‰ˆæœ¬
status: active

lineage:
  introduced_in: "v0.10"
  derived_from: "rule_r01_no_execution"  # æŒ‡å‘çˆ¶ç‰ˆæœ¬
  supersedes: ["rule_r01_no_execution"]  # æ›¿ä»£æ—§ç‰ˆæœ¬

# ... ä¿®æ”¹åçš„è§„åˆ™é€»è¾‘
```

---

## ğŸ“š å‚è€ƒèµ„æ–™

- **Schema**: `agentos/schemas/content/rule.schema.json`
- **ç¤ºä¾‹**: `docs/content/rules/p0/rule_r01_no_execution.yaml`ï¼ˆæ ‡å‡†æ¨¡æ¿ï¼‰
- **Validator**: `agentos/core/gates/validate_rule_redlines.py`
- **README**: `docs/content/rules/README.md`
- **Catalog**: `docs/content/rules/catalog.md`

---

## ğŸ’¬ è´¡çŒ®æµç¨‹

1. Fork ä»“åº“
2. åˆ›å»ºæ–° rule YAML æ–‡ä»¶
3. è¿è¡Œæœ¬åœ°æµ‹è¯•ï¼ˆschema + çº¢çº¿ï¼‰
4. æäº¤ PRï¼Œæ ‡é¢˜ï¼š`feat(rules): add R<nn> <short_title>`
5. ç­‰å¾… CI é€šè¿‡ï¼ˆ6 ä¸ª gatesï¼‰
6. ç­‰å¾… Code Review æ‰¹å‡†

---

**ç‰ˆæœ¬**: v0.9.0  
**ç»´æŠ¤**: AgentOS Team  
**æœ€åæ›´æ–°**: 2026-01-25
